src/staldates/joystick.py

```diff
@@ -188,14 +192,54 @@ def _update_camera(self):
             pass
 
 
-if __name__ == "__main__":
-    dev_str = "/dev/input/js1"
-    js = Joystick(dev_str)
-    cja = CameraJoystickAdapter(js)
+JOYSTICK_MAX = 32767
+JOYSTICK_HALF = JOYSTICK_MAX / 2
+
 
-    cam = VISCACamera("Camera 1", "/dev/ttyUSB0", 1)
-    cam.initialise()
+class SensitivityPrefsCameraJoystickAdapter(CameraJoystickAdapter):
+    def update_preferences(self):
+        CameraJoystickAdapter.update_preferences(self)
+        self._set_speed_params()
+
+    def set_camera(self, camera):
+        super(SensitivityPrefsCameraJoystickAdapter, self).set_camera(camera)
+        self._set_speed_params()
+
+    def _set_speed_params(self):
+        if self._camera:
+            self.max_pan = self.camera.maxPanSpeed
+            self.max_tilt = self.camera.maxTiltSpeed
+            self.min_zoom = self.camera.minZoomSpeed
+            self.max_zoom = self.camera.maxZoomSpeed
+        else:
+            self.max_pan = 0x18
+            self.max_tilt = 0x14
+            self.min_zoom = 2
+            self.max_zoom = 7
+
+        self.pan_sensitivity = Preferences.get('joystick.sensitivity.pan', 0.5)
+        self.tilt_sensitivity = Preferences.get('joystick.sensitivity.tilt', 0.5)
+        self.zoom_sensitivity = Preferences.get('joystick.sensitivity.zoom', 0.5)
+
+    def _interp(self, raw, max_value, sensitivity):
+        sensitivity_value = math.ceil(max_value * sensitivity)
+        if raw <= JOYSTICK_HALF:
+            return max(1, int(math.ceil(sensitivity_value * raw / JOYSTICK_HALF)))
+        else:
+            return max(1, int(sensitivity_value + math.ceil((2 * (raw - JOYSTICK_HALF) * (max_value - sensitivity_value) / JOYSTICK_MAX))))
+
+    def map_pan(self, axis):
+        return self._interp(abs(axis), self.max_pan, self.pan_sensitivity)
+
+    def map_tilt(self, axis):
+        return self._interp(abs(axis), self.max_tilt, self.tilt_sensitivity)
+
+    def map_zoom(self, axis):
+        return max(self.min_zoom, self._interp(abs(axis), self.max_zoom, self.zoom_sensitivity))
+
+
+if __name__ == "__main__":
```

Comment:
This looks like ad-hoc test code. It should probably be removed, and ideally replaced with proper unit tests.